src/analysis/analytic/uniqueness.lean

```diff
@@ -0,0 +1,106 @@
+theorem eq_on_zero_of_preconnected_of_eventually_eq_zero_aux [complete_space F]
+  {f : E → F} {U : set E} (hf : analytic_on 𝕜 f U) (hU : is_preconnected U) {w : E}
+  (hw : w ∈ U) (hfw : f =ᶠ[𝓝 w] 0) : eq_on f 0 U :=
+begin
+  /- Let `u` be the set of points around which `f` vanishes. It is clearly open. We have to show
+  that its limit points in `U` still belong to it, from which the inclusion `U ⊆ u` will follow
+  by connectedness.
+  Take a limit point `x`, then a ball `B (x, r)` on which it has a power series expansion, and
+  then `y ∈ B (x, r/2) ∩ u`. Then `f` has a power series expansion on `B (y, r/2)` as it is
+  contained in `B (x, r)`. All the coefficients in this series expansion vanish, as `f` is zero on a
+  neighborhood of `y`. Therefore, `f` is zero on `B (y, r/2)`. As this ball contains `x`, it follows
+  that `f` vanishes on a neighborhood of `x`, proving the claim. -/
+  let u := {x | f =ᶠ[𝓝 x] 0},
+  have main : closure u ∩ U ⊆ u,
+  { rintros x ⟨xu, xU⟩,
+    rcases hf x xU with ⟨p, r, hp⟩,
+    obtain ⟨y, yu, hxy⟩ : ∃ y ∈ u, edist x y < r / 2,
+      from emetric.mem_closure_iff.1 xu (r / 2) (ennreal.half_pos hp.r_pos.ne'),
+    let q := p.change_origin (y - x),
+    have has_series : has_fpower_series_on_ball f q y (r / 2),
+    { have A : (∥y - x∥₊ : ℝ≥0∞) < r / 2, by rwa [edist_comm, edist_eq_coe_nnnorm_sub] at hxy,
+      have := hp.change_origin (A.trans_le ennreal.half_le_self),
+      simp only [add_sub_cancel'_right] at this,
+      apply this.mono (ennreal.half_pos hp.r_pos.ne'),
+      apply ennreal.le_sub_of_add_le_left ennreal.coe_ne_top,
+      apply (add_le_add (A.le) (le_refl (r / 2))).trans (le_of_eq _),
+      exact ennreal.add_halves _ },
+    have M : emetric.ball y (r / 2) ∈ 𝓝 x, from emetric.is_open_ball.mem_nhds hxy,
+    filter_upwards [M] with z hz,
+    have A : has_sum (λ (n : ℕ), q n (λ (i : fin n), z - y)) (f z) := has_series.has_sum_sub hz,
+    have B : has_sum (λ (n : ℕ), q n (λ (i : fin n), z - y)) (0),
+    { have : has_fpower_series_at 0 q y, from has_series.has_fpower_series_at.congr yu,
+      convert has_sum_zero,
+      ext n,
+      exact this.apply_eq_zero n _ },
+    exact has_sum.unique A B },
+  have Uu : U ⊆ u, from
+    hU.subset_of_closure_inter_subset is_open_set_of_eventually_nhds ⟨w, hw, hfw⟩ main,
+  assume z hz,
+  simpa using mem_of_mem_nhds (Uu hz),
+end
+
+/-- The *identity principle* for analytic functions: If an analytic function vanishes in a whole
+neighborhood of a point `z₀`, then it is uniformly zero along a connected set. For a one-dimensional
+version assuming only that the function vanishes at some points arbitrarily close to `z₀`, see
+`eq_on_zero_of_preconnected_of_frequently_eq_zero`. -/
+theorem eq_on_zero_of_preconnected_of_eventually_eq_zero
+  {f : E → F} {U : set E} (hf : analytic_on 𝕜 f U) (hU : is_preconnected U) {w : E}
+  (hw : w ∈ U) (hfw : f =ᶠ[𝓝 w] 0) :
+  eq_on f 0 U :=
+begin
+  let F' := uniform_space.completion F,
```

Comment:
Can this space completion argument be moved to `has_fpower_series_on_ball.change_origin` instead? This seems to be the only location where completeness is needed?

Reply:
The problem is that `change_origin` is not true without completeness, as the defining series might not converge.

Reply:
I'm not used to thinking about these things ... You mean that there is a series with coefficients in `F`, it converges in the completion of `F` by applying `change_origin` and some coercion, but we know that the limit is in `F` because it is the value of the initial function which takes values in `F`, and still this does not imply convergence in `F`?

Reply:
Yes. Each coefficient of the expansion at the new origin is given by a series of multilinear operators with summable norms. There is only a guarantee that this series converges to a well-defined operator when the target space is complete, so when it is not complete I can't even define the coefficients (although I agree with you that this would be super-weird). In any case, completeness will be a prerequisite in all serious applications, so getting rid of the completeness assumption here is not important at all (but still if it's possible we should do it...)

Reply:
Ah ok, so the coefficients of the new series are themselves in the completion. That's a bit annoying.

Is it true that if an analytic function takes values in a subspace then so do the coefficients of the series? As a generalization of analytic continuation (where the subspace is {0}) it sounds like it would perhaps be plausible, but OTOH the only thing I see how to prove is that the coefficients would be in the completion of the subspace, which would be kind of moral but would not help at all.

Reply:
I don't think it's true that the coefficients belong to the subspace generated by the values of the functions. And this is why completeness is necessary to say anything nontrivial in general.

Here is a funny example. Let `f` from `ℂ` (or rather the open unit disk) to `ℓ^1(ℂ)` map `z` to `(1, z, z^2, z^3, ...)`. Then the derivative at `0` is `(0, 1, 0, 0, ...)`, and it does not belong to the subspace spanned by the values of the `f (z)` (use the fact that Vandermonde determinants are nonzero to see that, if a linear combination of the `f (z_i)` vanishes after some point, then all the coefficients in the linear combination are zero except that of `f (0)`).

Reply:
Ah, that's a nice example. OTOH you do have to put `(0, 1, 0, ...)` in the space in which the coefficients of the series at `0` are chosen so if I understand it correctly this `f` is not `analytic_at 0` as a function from C to the span of its values.

Ah but it is `has_fpower_series_on_ball` of radius `1` as a function from C to the span of its values and sequences with finite supports while being `analytic_at` nowhere else with this space as codomain. OK, I think I am slowly getting it.